syntax = "proto3";

option go_package = "securechat-server/grpc";
package main;

service ClientServerComms {
	rpc SignUp(SignUpRequest) returns (SignUpResponse);
	rpc SignUpChallengeResponse(SignUpChallengeResponseRequest) returns (SignUpChallengeResponseResponse);
	rpc Login(LoginRequest) returns (LoginResponse);
	rpc FindUser(FindUserRequest) returns (FindUserResponse);
}

message SignUpRequest{
	string username = 1;
	bytes publicKeyLogin = 2;
	bytes publicKeyChat = 3;
}

message SignUpResponse{
	bytes challenge = 1;
}

message SignUpChallengeResponseRequest{
	bytes challenge_response = 1;
}

message SignUpChallengeResponseResponse{
	// assume no error means success
}

message LoginRequest{
	reserved 3;
	string username = 1;
	fixed32 address = 2;
	string digitalSignatureUsername = 4;
	bytes digitalSignatureSignature = 5;
}

message LoginResponse{
	// assume no error means success
}

message FindUserRequest{
	reserved 2;
	string username = 1;
	string digitalSignatureUsername = 3;
	bytes digitalSignatureSignature = 4;
}

message FindUserResponse{
	string username = 1;
	fixed32 address = 2;
	bytes publicKeyChat = 3;
}